src/System.Private.CoreLib/shared/System/Diagnostics/Tracing/CounterGroup.cs

```diff
@@ -205,48 +236,54 @@ private void OnTimer(object? state)
                         counter.WritePayload((float)elapsed.TotalSeconds, _pollingIntervalInMilliseconds);
                     }
                     _timeStampSinceCollectionStarted = now;
-                }
-                else
-                {
-                    DisposeTimer();
+
+                    do {
+                        _nextPollingTimeStamp += new TimeSpan(0, 0, 0, 0, _pollingIntervalInMilliseconds);
+                    } while(_nextPollingTimeStamp <= now);
                 }
             }
         }
 
-        #region PCL timer hack
 
-#if ES_BUILD_PCL
-        internal delegate void TimerCallback(object state);
 
-        internal sealed class Timer : CancellationTokenSource, IDisposable
-        {
-            private int _period;
-            private TimerCallback _callback;
-            private object _state;
-
-            internal Timer(TimerCallback callback, object state, int dueTime, int period)
-            {
-                _callback = callback;
-                _state = state;
-                _period = period;
-                Schedule(dueTime);
-            }
-
-            private void Schedule(int dueTime)
-            {
-                Task.Delay(dueTime, Token).ContinueWith(OnTimer, null, CancellationToken.None, TaskContinuationOptions.ExecuteSynchronously | TaskContinuationOptions.OnlyOnRanToCompletion, TaskScheduler.Default);
-            }
+        private static Thread? s_pollingThread;
+        private static int s_sleepDurationInMilliseconds;
+        private static ManualResetEvent? s_pollingThreadEvent;
+        private static List<CounterGroup>? s_counterGroupEnabledList;
 
-            private void OnTimer(Task t, object s)
+        private static void PollForValues()
+        {
+            while (true)
             {
-                Schedule(_period);
-                _callback(_state);
+                lock (s_counterGroupLock)
+                {
+                    if (s_counterGroupEnabledList != null)
+                    {
+                        foreach (CounterGroup counterGroup in s_counterGroupEnabledList)
+                        {
+                            DateTime now = DateTime.UtcNow;
+                            if (counterGroup._nextPollingTimeStamp < now + new TimeSpan(0, 0, 0, 0, 1))
+                            {
+                                counterGroup.OnTimer();
+                            }
+
+                            int millisecondsTillNextPoll = (int)((counterGroup._nextPollingTimeStamp - now).TotalMilliseconds);
+                            millisecondsTillNextPoll = millisecondsTillNextPoll > 1 ? millisecondsTillNextPoll : 1;
+                            if (millisecondsTillNextPoll < s_sleepDurationInMilliseconds)
+                            {
+                                s_sleepDurationInMilliseconds = millisecondsTillNextPoll;
```

Comment:
s_sleepDurationInMilliseconds never grows which will likely lead to this loop waking up more frequently than it needs to.